README
Module  : @TITLE@
Version : @VERSION@
Author  : Thomas Seifer <ts77@phorum.org>
          Maurice Makaay <maurice@phorum.org>

@DESCRIPTION@


Requirements:
-------------

To get this module to work, you will need to install the Sphinx system
on your server. This will normally not be possible in a shared hosting
environment (because most shared hosting providers won't allow you to
install extra third party software) and no, there is no workaround for that.

You will also need to be able to run scheduled jobs (a.k.a. "cron jobs"
on UNIX system) on the server for updating the search indexes periodically.


Installation:
-------------

- First, you will have to install the Sphinx software. You can either
  install this from some binary package or you can build it yourself from
  the source package. Which way you choose, is up to you. For more
  information on Sphinx and installing it, see http://www.sphinxsearch.com/

- After installing Sphinx, you have to generate the configuration file
  for Sphinx. A script is provided to help you with this task. Run this
  script on the command prompt:

    $ cd /path/to/your/phorum/mods/sphinx_search
    $ php install_helper.php

  Go over the settings and change them to your likings. The most important
  one is the DATAPATH setting. The directory that you enter here should be
  writable for the user that is going to index the database and the user
  that is going to run the Sphinx "searchd" program. After you are done
  with the settings, let the script generate the configuration file.
  It will tell you in where it has stored the generated file.

  If you choose to enable USE_DELTA_UPDATES, an additional table is
  required in the Phorum database. The install helper script can
  automatically create this table for you.

- Move the generated configuration file to a useful location on the server.
  You might for example use "/etc/sphinx.conf" on a UNIX system.

- Before starting Sphinx for the first time, you need to build an initial
  set of index files. See the Sphinx documentation for the "indexer"
  program for more information on the indexing process. On my Linux system,
  I used the following command:

    $ /opt/sphinx/bin/indexer --config /etc/sphinx.conf --all

- Now, the Sphinx "searchd" program can be started up. This is the program
  that will accept search queries and will return the search results
  to Phorum. I used the following command:

    $ /opt/sphinx/bin/searchd --config /etc/sphinx.conf

- At this point, you should be able to use the Sphinx search module in
  Phorum. First go to Phorum's admin "Modules" screen and enable the
  Sphinx search module. Then try a search query on your Phorum install.


============================================================================
============================================================================

some caveats:
  - data is stored in both phorum_search-table (not needed here) and in the
    sphinx filesystem - this doubles the sizes needed for search even though
    the search-table is not needed. You can clear it out every now and then.
  - search-data is not updated in real-time - sphinx reindexes the data if
    the indexer is called for the index, which does a full reindexing of
    the sphinx index. There is a workaround available in sphinx
    http://sphinxsearch.com/doc.html#live-updates which works essentially
    with two indexes. A large one and smaller one with only the deltas
    to the full index and which can be reindexed therefore more often,
    but I don't use that method yet. Still, its not realtime. For me the
    full indexing of 350,000 messages took around 5 minutes. I'd run the
    full indexing around once a day with those times.

Before you can run a search through sphinx you will need to setup sphinx
correctly and do at least one full indexing of your phorum-data.
Try the searching on the command-line using search from the sphinx install
first. Also you will need to start searchd from sphinx so that the mod can
connect to it for searching.



Changelog
Changelog for @PACKAGE@
----------------------------------------------------------------------

v0.9.5 - Okt 23, 2007

    - Added Turkish language file. 
      Thanks to Serdar of paticik.com.

v0.9.4 - Okt 09, 2007

    - Brought all code up to date and cleaned up the configuration
      template a bit.

    - Added some example startup and indexing scripts.

    - Added an option to search for body+subject, body or subject.

    - Built in some work arounds for a few problems that are in the
      Sphinx extended query parser.

v0.9.3 - Apr 25, 2007

    - Added a settings screen to be able to configure the settings
      for the module from the admin interface.

    - Made the results limit configurable.

    - Wrote an install helper script that can be used for easily
      configuring Sphinx and Phorum.

v0.9.2 - Apr 24, 2007 by Thomas Seifert (thomas@phorum.org)

    - Limited results to 1000.
    - Some security for the forum_id.

v0.9.1 - Dec 02, 2006 by Thomas Seifert (thomas@phorum.org)

    - Updated to sphinx 0.9.7, added delta indexes and a host-ip

v0.9 - Oct 13, 2006 by Maurice Makaay (maurice@phorum.org)

    - Return search control to the Phorum search script in case an
      unhandled type of search is encountered (e.g. the USER_ID search
      type for Phorum 5.2).

v0.8 - Sep 13, 2006 by Maurice Makaay (maurice@phorum.org)

    - little hack in SetGroups of the sphinxapi.php script to not deny
      group 0 (is_int(0) is false, but we can have 0 for announcements)
    - if no results are found, return empty results immediately (before,
      no results would trigger undefined index warnings)
    - changed example config in README to use settings inheritance

v0.7 - Sep 07, 2006

    - received some more testing and hence ...
    - fix for date-order

v0.5 - Jul 13, 2006

    - first implementation, without admin interface



defaults.php
<?php

if (! isset($GLOBALS["PHORUM"]["mod_sphinx_search"])) {
    $GLOBALS["PHORUM"]["mod_sphinx_search"] = array();
}

if (! isset($GLOBALS["PHORUM"]["mod_sphinx_search"]["sphinx_host"])) {
    $GLOBALS["PHORUM"]["mod_sphinx_search"]["sphinx_host"] = "localhost";
}

if (! isset($GLOBALS["PHORUM"]["mod_sphinx_search"]["sphinx_port"])) {
    $GLOBALS["PHORUM"]["mod_sphinx_search"]["sphinx_port"] = 3312;
}

if (! isset($GLOBALS["PHORUM"]["mod_sphinx_search"]["message_index"])) {
    $GLOBALS["PHORUM"]["mod_sphinx_search"]["message_index"] =
        "phorum5_msg_d phorum5_msg";
}

if (! isset($GLOBALS["PHORUM"]["mod_sphinx_search"]["author_index"])) {
    $GLOBALS["PHORUM"]["mod_sphinx_search"]["author_index"] =
        "phorum5_author_d phorum5_author";
}

if (! isset($GLOBALS["PHORUM"]["mod_sphinx_search"]["max_search_results"])) {
    $GLOBALS["PHORUM"]["mod_sphinx_search"]["max_search_results"] = 1000;
}

?>


settings.php
<?php

// Check if we are loaded from the Phorum admin code.
// Direct access to this file is not allowed.
if (! defined("PHORUM_ADMIN")) return;

require_once("./mods/sphinx_search/defaults.php");

// For shorter writing.
$settings =& $GLOBALS["PHORUM"]["mod_sphinx_search"];

if (count($_POST))
{
    // Check for errors in the settings.

    $errors = array();

    $_POST["sphinx_host"] = trim($_POST["sphinx_host"]);
    if (!preg_match('/^[\w\.-]+$/', $_POST["sphinx_host"])) {
        $errors[] = "The Sphinx host field should contain a valid host name.";
    }
    $settings["sphinx_host"] = $_POST["sphinx_host"];

    if (!is_numeric($_POST["sphinx_port"])) {
        $errors[] = "The Sphinx port field should contain a numerical value.";
    }
    $settings["sphinx_port"] = (int) $_POST["sphinx_port"];

    $settings["message_index"]  = trim($_POST["message_index"]);
    $settings["author_index"]   = trim($_POST["author_index"]);

    if (!is_numeric($_POST["max_search_results"])) {
        $errors[] = "The maximum search results field should contain a numerical value.";
    }
    $settings["max_search_results"] = (int) $_POST["max_search_results"];

    if (count($errors)) {
        phorum_admin_error(implode("<br>", $errors));
    } else {
        phorum_db_update_settings(array("mod_sphinx_search" => $settings));
        phorum_admin_okmsg("The settings were successfully saved.");
    }
}

// Create the settings form.
include_once "./include/admin/PhorumInputForm.php";
$frm = new PhorumInputForm ("", "post", "Save settings");
$frm->hidden("module", "modsettings");
$frm->hidden("mod", "sphinx_search");
$frm->hidden("sp_action", "settings");

$frm->addbreak("Module Settings");

$frm->addsubbreak("Sphinx server configuration");

$row = $frm->addrow("Sphinx server IP address or hostname", $frm->text_box("sphinx_host", $settings["sphinx_host"], 50));

$row = $frm->addrow("Sphinx server port", $frm->text_box("sphinx_port", $settings["sphinx_port"], 10));

$row = $frm->addrow("Message index(es) to use", $frm->text_box("message_index", $settings["message_index"], 50));

$row = $frm->addrow("Author index(es) to use", $frm->text_box("author_index", $settings["author_index"], 50));

$row = $frm->addrow("Maximum number of search results", $frm->text_box("max_search_results", $settings["max_search_results"], 10));

$frm->show();

?>


sphinx_search.php
<?php

require_once("./mods/sphinx_search/defaults.php");

function mod_sphinx_change_defaults($args)
{
    if(isset($args["page"])) {
        return $args;
    }

    $args['match_dates'] = "0";

    return $args;
}

function mod_sphinx_search_redirect($url_parameters)
{
    // Add the extra form fields to the primary search redirect URL.
    // Apply some defaults if the settings are not available.

    if (isset($_GET["sphinx_match_fields"])) {
        $url_parameters[] = 'sphinx_match_fields='.urlencode($_GET["sphinx_match_fields"]);
    } else {
        $url_parameters[] = 'sphinx_match_fields=BOTH';
    }

    if (isset($_GET["sphinx_match_threads"])) {
        $url_parameters[] = 'sphinx_match_threads='.(int)$_GET["sphinx_match_threads"];
    } else {
        $url_parameters[] = 'sphinx_match_threads=0';
    }

    return $url_parameters;
}

function mod_sphinx_do_search($data_array)
{
    $PHORUM=$GLOBALS['PHORUM'];

    $settings = $PHORUM["mod_sphinx_search"];
    settype($settings["sphinx_port"], "int");

    // Initialize the Sphinx client.
    include './mods/sphinx_search/sphinxapi.php';
    $sphinx = new SphinxClient();
    $sphinx->SetServer($settings["sphinx_host"], $settings["sphinx_port"]);
    $sphinx->SetMatchMode(SPH_MATCH_EXTENDED);

    // Set the limits for paging.
    $sphinx->SetLimits($data_array['offset'], $data_array['length']);

    // In this array, we'll build the Sphinx queries to run.
    $match_strs = array();

    // Handle search on body or subject terms.
    $search = trim($data_array['search']);

    if($search != '')
    {
        $tokens = array();

        // Prepare searching for an exact phrase.
        if ($data_array['match_type'] == "PHRASE")
        {
            $tokens[] = '"'. str_replace('"', '', $search) .'"';
        }
        // Prepare searching for separate query terms.
        else
        {
            // Surround the query with spaces so matching is easier.
            $search = " $search ";

            // Pull all grouped terms from the query,
            // e.g. (nano mini)
            $paren_terms = array();
            if ( strstr($search, '(') ) {
                preg_match_all('/ ([+\-~]*\(.+?\)) /', $search, $m);
                $search = preg_replace('/ [+\-~]*\(.+?\) /', ' ', $search);
                $paren_terms = $m[1];
            }

            // Pull all double quoted terms from the query,
            // e.g. "iMac DV" or -"iMac DV"
            $quoted_terms = array();
            if ( strstr($search, '"') ) {
                preg_match_all('/ ([+\-~]*".+?") /', $search, $m);
                $search = preg_replace('/ [+\-~]*".+?" /', ' ', $search);
                $quoted_terms = $m[1];
            }

            // Finally, pull the rest words from the query.
            $norm_terms = preg_split("/\s+/", $search, 0, PREG_SPLIT_NO_EMPTY);

            // Replace + and ~ in the input search; no use for them in Sphinx.
            $norm_terms = preg_replace( '/^[+~]/', '', $norm_terms );

            // Merge all parts together.
            $tokens =  array_merge( $quoted_terms, $paren_terms, $norm_terms );
        }

        // What fields to search on. This can be used to restrict the
        // search to body or subject only.
        $search_fields = array('@body', '@subject');
        if (isset($PHORUM["args"]["sphinx_match_fields"])) {
            if ($PHORUM["args"]["sphinx_match_fields"] == 'BODY') {
                $search_fields = array('@body');
            } elseif ($PHORUM["args"]["sphinx_match_fields"] == 'SUBJECT') {
                $search_fields = array('@subject');
            }

            $GLOBALS["PHORUM"]["DATA"]["SEARCH"]["sphinx_match_fields"] =
                htmlspecialchars($PHORUM["args"]["sphinx_match_fields"]);
        }

        // putting together the actual search
        if ($data_array['match_type'] == 'ALL' ||
            $data_array['match_type'] == 'ANY')
        {
            $subtokens = array();
            $negates   = '';
            foreach ($tokens as $token) {
                if (substr($token, 0, 1) == '-') {
                    foreach ($search_fields as $field) {
                        $negates .= "$field $token ";
                    }
                } else {
                    $parts = array();
                    foreach ($search_fields as $field) {
                        $parts[] = "$field $token";
                    }
                    $subtokens[] = implode(' | ', $parts);
                }
            }

            $glue = $data_array['match_type'] == 'ALL' ? ' & ' : ' | ';

            $match_str = $negates . implode($glue, $subtokens);
        }
        elseif ($data_array['match_type'] == 'PHRASE')
        {
            $subtokens = array();
            foreach ($search_fields as $field) {
                $subtokens[] = "$field " . $tokens[0];
            }
            $match_str = implode(' | ', $subtokens);
        }
        else
        {
            // Return search control to Phorum in case the search type
            // isn't handled by the module.
            return $data_array;
        }

        $match_strs[] = $match_str;
    }

    // add author matching
    $author = trim($data_array['author']);
    if ($author != '') {
        // The USER_ID match_type is not handled by the Sphinx backend.
        if ($data_array['match_type'] == 'USER_ID') {
            return $data_array;
        } else {
            $match_strs[] ="@author ".$author;
        }
    }

    $match_str = implode(" & ",$match_strs);

    // set the timeframe to search
    if($data_array['match_dates'] > 0)
    {
        $min_ts = time() - 86400 * $data_array['match_dates'];
        $max_ts = time();
        $sphinx->SetFilterRange("datestamp", $min_ts, $max_ts);
    }

    // add the forum(s) to search
    if ($data_array['match_forum'] == 'THISONE')
    {
        $forumid_clean = (int)$PHORUM['forum_id'];
        $sphinx->SetFilter ( "forum_id", array($forumid_clean) );
    }
    elseif(!empty($data_array['match_forum']))
    {
        // We have to check what forums they can read first.
        $allowed_forums = phorum_api_user_check_access(
            PHORUM_USER_ALLOW_READ,
            PHORUM_ACCESS_LIST
        );

        // Prepare forum_id restriction.
        $match_forum_arr = explode(",", $data_array['match_forum']);
        foreach ($match_forum_arr as $forum_id)
        {
            if ($forum_id=="ALL") {
                $search_forums = $allowed_forums;
                break;
            }
            if (isset($allowed_forums[$forum_id])){
                $search_forums[] = (int)$forum_id;
            }
        }

        // If they are not allowed to search any forums or if the currently
        // active forum is not readable, then return empty handed.
        if(empty($allowed_forums) || ($PHORUM['forum_id']>0 &&
           !in_array($PHORUM['forum_id'], $allowed_forums)))
        {
            $data_array['results']  = array();
            $data_array['totals']   = 0;
            $data_array['continue'] = 0;
            $data_array['raw_body'] = 1;

            return $data_array;
        }

        $sphinx->SetFilter ( "forum_id", $search_forums );
    }

    // Set the sort-mode.
    $sphinx->SetSortMode(SPH_SORT_ATTR_DESC, 'datestamp');

    // Handle thread-grouping.
    if (!empty($PHORUM['args']['sphinx_match_threads']) ||
        $data_array['match_threads']) {
        $sphinx->SetGroupBy('thread', SPH_GROUPBY_ATTR);
    }
    if (!empty($PHORUM['args']['sphinx_match_threads'])) {
        $GLOBALS["PHORUM"]["DATA"]["SEARCH"]["sphinx_match_threads"] =
            htmlspecialchars($PHORUM["args"]["sphinx_match_threads"]);
    }

    // Run the actual query through Sphinx.
    $results = $sphinx->Query($match_str, $settings['message_index']);

    // Show a warning on screen if there was a Sphinx problem.
    $error = $sphinx->GetLastError();
    if (!empty($error)) {
        $GLOBALS['PHORUM']['DATA']['GLOBAL_ERROR'] =
            'Sphinx search engine error: ' .
            htmlspecialchars($error);
    }

    // if no messages were found, then return empty handed.
    if (empty($results["matches"]))
    {
        $data_array['results']  = array();
        $data_array['totals']   = 0;
        $data_array['continue'] = 0;
        $data_array['raw_body'] = 1;

        return $data_array;
    }

    // Retrieve the messages that we found from the database.
    if ((!empty($PHORUM['args']['sphinx_match_threads']) &&
        $PHORUM['args']['sphinx_match_threads'] == 1) ||
        (empty($PHORUM['args']['sphinx_match_threads']) &&
         !empty($data_array['match_threads'])))
    {
        $message_ids = array();
        foreach ($results['matches'] as $id => $data) {
            $message_ids[] = $data['attrs']['thread'];
        }
    } else {
        $message_ids = array_keys($results['matches']);
    }
    $found_messages = phorum_db_get_message($message_ids, 'message_id', TRUE);

    // Sort them in reverse order of the message_id, to automagically
    // sort them by date desc this way.
    krsort($found_messages);
    reset($found_messages);

    // Prepare the array for letting Sphinx build highlighted excerpts.
    require_once('./include/format_functions.php');
    $found_messages = phorum_format_messages($found_messages);
    $docs=array();
    foreach($found_messages as $id => $data)
    {
        $found_messages[$id]['subject'] = strip_tags($data['subject']);
        $docs[] = phorum_strip_body($data['body']);
    }

    $words = implode(" ",array_keys($results['words']));

    $opts = array('chunk_separator'=>' [...] ');

    // The excerpts index is mainly used for charsets and mapping tables.
    // It always only needs one index. We extract one index from the
    // message index.
    $excerpts_indexs = explode(" ", $settings["message_index"]);
    $excerpts_index  = $excerpts_indexs[0];

    // build highlighted excerpts
    $highlighted = $sphinx->BuildExcerpts($docs,$excerpts_index,$words,$opts);

    // Show a warning on screen if there was a Sphinx problem.
    $error = $sphinx->GetLastError();
    if (!empty($error)) {
        $GLOBALS['PHORUM']['DATA']['GLOBAL_ERROR'] =
            'Sphinx search engine error: ' .
            htmlspecialchars($error);
    }

    $cnt=0;
    foreach($found_messages as $id => $content) {
    // foreach($found_messages as $id => $content) {
        $found_messages[$id]['short_body'] = $highlighted[$cnt];
        $cnt++;
    }

    $data_array['results']=$found_messages;
    // we need the total results
    $data_array['totals']=$results['total_found'];

    if($data_array['totals'] > $settings["max_search_results"]) {
        $data_array['totals'] = $settings["max_search_results"];
    }

    // don't run the default search
    $data_array['continue']=0;
    // tell it to leave the body alone
    $data_array['raw_body']=1;

    return $data_array;
}

function mod_sphinx_search_output($template)
{
    // Apply some defaults for the Sphinx module specific fields.
    if (empty($GLOBALS["PHORUM"]["DATA"]["SEARCH"]["sphinx_match_fields"])) {
        $GLOBALS["PHORUM"]["DATA"]["SEARCH"]["sphinx_match_fields"] = 'BOTH';
    }
    if (empty($GLOBALS["PHORUM"]["DATA"]["SEARCH"]["sphinx_match_threads"])) {
        $GLOBALS["PHORUM"]["DATA"]["SEARCH"]["sphinx_match_threads"] = 0;
    }

    // Use our own search template, so we can add a feature for
    // searching only in the header or body.
    return 'sphinx_search::search';
}

?>


install_helper.php
<?php

// if we are running in the webserver, bail out
if (isset($_SERVER["REMOTE_ADDR"])) {
    trigger_error(
        "This script cannot be run from a browser.\n",
        E_USER_ERROR
    );
}

define("phorum_page", "sphinx_install_helper");
define("PHORUM_ADMIN", 1);

// I guess the phorum-directory should be two levels up.
$PHORUM_DIRECTORY = dirname(__FILE__) . "/../../";

// change directory to the main-dir so we can use common.php
if(file_exists($PHORUM_DIRECTORY."/common.php")) {
    chdir($PHORUM_DIRECTORY);
    if (!is_readable("./common.php")) {
        fprintf(STDERR,
            "Unable to read common.php from directory $PHORUM_DIRECTORY\n");
        exit(1);
    }
} else {
    fprintf(STDERR,
        "Unable to find Phorum file \"common.php\".\n" .
        "Please check the \$PHORUM_DIRECTORY in " . basename(__FILE__) ."\n");
    exit(1);
}

require_once("common.php");

// Check the db configuration. We only do Sphinx for MySQL, so check if
// we have MySQL as our database and if it's fully configured.
if (!strstr($PHORUM["DBCONFIG"]["type"], "mysql")) {
    trigger_error(
        "It looks like you database type is not MySQL.\n" .
        "This module only runs with MySQL as the database backend.\n",
        E_USER_ERROR
    );
}
foreach (array("server", "name", "user", "password", "table_prefix") as $fld) {
    if (!isset($PHORUM["DBCONFIG"][$fld])) trigger_error(
        "The configuration field \"$fld\" is missing in your Phorum\n" .
        "database configuration include/db/config.php. Please fix this\n" .
        "and retry running this script.\n",
        E_USER_ERROR
    );
}

print "\nSphinx install helper\n" .
      "---------------------\n\n";

// This array holds the Sphinx configuration variables that we are using.
$options = array
(
    array(
        "DBHOST",
        $PHORUM["DBCONFIG"]["server"],
        "The hostname for connecting to the Phorum database server\n" .
        "\n" .
        "You should not have to change this value, since the default value\n".
        "is the actual value that Phorum is uses for accessing your Phorum\n".
        "database data."
    ),
    array(
        "DBPORT",
        3306,
        "The port at which the Phorum database server is running (default\n" .
        "is \"3306\" and normally you should not have to change that value)."
    ),
    array(
        "DBNAME",
        $PHORUM["DBCONFIG"]["name"],
        "The name of the Phorum database.\n" .
        "\n" .
        "You should not have to change this value, since the default value\n".
        "is the actual value that Phorum is uses for accessing your Phorum\n".
        "database data."
    ),
    array (
        "DBUSER",
        $PHORUM["DBCONFIG"]["user"],
        "The username for accessing the Phorum database.\n" .
        "\n" .
        "You should not have to change this value, since the default value\n".
        "is the actual value that Phorum is uses for accessing your Phorum\n".
        "database data."
    ),
    array (
        "DBPASS",
        $PHORUM["DBCONFIG"]["password"],
        "The password for the Phorum database.\n" .
        "\n" .
        "You should not have to change this value, since the default value\n".
        "is the actual value that Phorum is uses for accessing your Phorum\n".
        "database data."
    ),
    array (
        "DBPREFIX",
        $PHORUM["DBCONFIG"]["table_prefix"],
        "The database prefix for the Phorum database.\n" .
        "\n" .
        "You should not have to change this value, since the default value\n".
        "is the actual value that Phorum is uses for accessing your Phorum\n".
        "database data."
    ),
    array (
        "MESSAGE_TABLE",
        isset($PHORUM["message_table"])
        ? $PHORUM["message_table"] : "",
        "This is the name of the Phorum database table in which the forum\n" .
        "messages are stored.\n" .
        "\n" .
        "You should not have to change this value, since the default value\n".
        "is the actual value that Phorum is uses for accessing your Phorum\n".
        "database data."
    ),
    array (
        "DATAPATH",
        "/var/sphinx",
        "The path to the directory where you will store the Sphinx data\n" .
        "files. This directory should be writable for the user which is\n" .
        "running the Sphinx software and readable for the user which is\n" .
        "running the webserver software.\n" .
        "\n" .
        "Note: This path does not determine where the generated Sphinx\n" .
        "configuration file will be written. You will be able to enter the\n" .
        "location for writing that file, after choosing the \"g\"enerate \n" .
        "menu option.",
        TRUE
    ),
    array (
        "USE_DELTA_UPDATES",
        0,
        "If you set this option to a true value (e.g. \"1\"), then the\n" .
        "configuration file will have support for doing \"delta updates\"\n" .
        "for the search data. This means that two different indexes will\n" .
        "be used by the Sphinx system. One index will be updated on a\n" .
        "slow schedule (like twice a day). This index will hold all data\n" .
        "for messages that were posted up to the time of indexing. The\n" .
        "other index will be updated on a really regular base (like each\n" .
        "15 minutes or even more regular). This index will hold all data\n" .
        "for messages that were posted after the last full index update.\n" .
        "\n" .
        "Note: for running the delta updates schedule, you will need to run\n".
        "MySQL server version 4.1 or higher. It will not work on older\n" .
        "versions of MySQL because of lacking features.",
        TRUE
    ),
    array (
        "MAX_SEARCH_RESULTS",
        1000,
        "This is the maximum number of search results that the Sphinx\n" .
        "search engine will return to Phorum (default is \"1000\").\n"
    ),
    array (
        "DATA_PREFIX",
        $PHORUM["DBCONFIG"]["table_prefix"],
        "This prefix is used as a prefix for all the Sphinx sources,\n" .
        "indexes and data files. You can choose anything here you like,\n" .
        "as long as this prefix is unique for the DATAPATH setting. If you\n" .
        "are storing multiple Sphinx search data sets in the same DATAPATH,\n".
        "then you would need to differentiate the DATA_PREFIXes to prevent\n" .
        "file name collisions from happening.\n" .
        "\n" .
        "If you don't know what value to use, then just use the default"
    ),
    array (
        "SEARCHD_HOST",
        "localhost",
        "This is the hostname for connecting to the\n" .
        "Sphinx searchd server (default is \"localhost\")."
    ),
    array (
        "SEARCHD_PORT",
        "3312",
        "This is is the port number for connecting to the\n" .
        "Sphinx searchd server (default is \"3312\")."
    )
);

while (TRUE)
{
    print <<<INFO
Please select the configuration option that you want to change. The most
interesting features are marked with "***". When selecting an option, you will
get some more info about the purpose of that option. You can safely return
to this screen after doing so.


INFO;

    foreach ($options as $id => $option) {
        $flag = isset($option[3]) ? "***" : "";
        print ($id+1) . ") $option[0] = $option[1] $flag\n";
    }
    print "q) Quit from this program.\n";
    print "g) Generate configuration file.\n";

    print "\n> ";
    $opt = trim(fgets(STDIN));
    if (((int)$opt > 0 && (int)$opt < count($options)))
    {
        $id = (int)$opt - 1;
        $option = $options[$id];

        print "\nOption: $option[0]\n";
        print str_repeat("-", 70) . "\n";
        print "$option[2]\n";
        print str_repeat("-", 70) . "\n";
        print "Enter new value or <ENTER> to accept the displayed current value.\n";
        print "\n";
        print $option[1] . "> ";
        $val = trim(fgets(STDIN));
        if ($val != '') {
            $options[$id][1] = $val;
        }
        print "\n";
    }
    elseif ($opt == 'q') {
        print "\nAborting ...\n\n";
        exit();
    }
    elseif ($opt == 'g') {
        print "\n";
        $errors = 0;
        foreach ($options as $option) {
            if ($option[1] === NULL || $option[1] === '') {
                print "Option \"$option[0]\" must be configured.\n";
                $errors = 1;
            }
        }
        if ($errors) {
            print "\n";
            sleep(1);
            continue;
        } else {
            break;
        }
    }
    else
    {
        print "\nIllegal option selected. Please try again ...\n\n";
        sleep(1);
        continue;
    }
}

foreach ($options as $option) {
    $PHORUM["DATA"][$option[0]] = $option[1];
}

$PHORUM["template"] = 'config_template';
ob_start();
include(phorum_get_template("sphinx_search::configuration"));
$config = ob_get_contents();
ob_end_clean();

$outfile = $PHORUM["cache"] . "/sphinx.conf-" . time();
require_once("./include/api/write_file.php");
phorum_api_write_file($outfile, $config);
print <<< CONFIGDONE
----------------------------------------------------------------------
A new configuration has been written to the location:
$outfile
----------------------------------------------------------------------

Look over the generated configuration file and see if you want to change
anything. After that, the first thing to do, would be to get Sphinx up and
running using this configuration file. For doing so, please refer to the
Sphinx documentation. The tasks that have to be taken care of are:

* Indexing your Phorum data, using the Sphinx "indexer" program.

* Running the Sphinx "searchd" program, which will let Phorum access the
  indexed search data.

* Setting up scheduled jobs for letting Sphinx update its indexes on a
  regular basis.

After setting up Sphinx, you will be ready to enable the Sphinx Search module.
You will have to configure that module from its settings page. If you wish,
this install script can update the module settings for you. Do you want
to run this update now?

CONFIGDONE;

while (TRUE)
{
    print "\nPlease enter \"yes\" or \"no\" > ";
    $yesno = trim(fgets(STDIN));
    if ($yesno == 'yes' || $yesno == 'no') { break; }
}

$prefix = $PHORUM["DATA"]["DATA_PREFIX"];

$mi = "{$prefix}_msg" .
      ($PHORUM["DATA"]["USE_DELTA_UPDATES"]?" {$prefix}_msg_delta":"");

$ai = "{$prefix}_author" .
      ($PHORUM["DATA"]["USE_DELTA_UPDATES"]?" {$prefix}_author_delta":"");

$settings = array (
    "sphinx_host"        => $PHORUM["DATA"]["SEARCHD_HOST"],
    "sphinx_port"        => $PHORUM["DATA"]["SEARCHD_PORT"],
    "message_index"      => $mi,
    "author_index"       => $ai,
    "max_search_results" => $PHORUM["DATA"]["MAX_SEARCH_RESULTS"],
);

if ($yesno == 'yes')
{
    phorum_db_update_settings(array("mod_sphinx_search" => $settings));

    print "\nThe settings for the Sphinx Search module have been updated.\n" .
          "You should only have to enable the module to get it working\n" .
          "with the Sphinx search engine.\n\n";
} else {
    print "\n" .
      "You will have to configure the settings for the module by\n" .
      "hand. Below, you will see the options that you will need to\n" .
      "use in the settings screen for this module, when running with\n" .
      "the generated Sphinx configuration file:\n\n" .
      "Sphinx IP address or hostname : {$settings["sphinx_host"]}\n" .
      "Sphinx port                   : {$settings["sphinx_port"]}\n" .
      "Message index(es) to use      : $mi\n" .
      "Author index(es) to use       : $ai\n" .
      "Maximum number of results     : {$settings["max_search_results"]}\n".
      "\n";
}

CONFIGDONE;

// No delta updates? Then we are done!
if (! $PHORUM["DATA"]["USE_DELTA_UPDATES"]) exit(0);

// Check if the incremental delta updates support table is available
// already. If not, then prompt for creating it now. Otherwise, we're done.
$incr_table = "{$PHORUM["DBCONFIG"]["table_prefix"]}_sphinx_counter";
$incr_recs = phorum_db_interact(
    DB_RETURN_ROWS,
    "DESC $incr_table",
    NULL,
    DB_MISSINGTABLEOK
);
if (!empty($incr_recs)) exit(0);


$sql = "CREATE TABLE $incr_table (\n" .
       "  counter_id int(11) unsigned NOT NULL default '0',\n" .
       "  `type` enum('author','message') NOT NULL default 'message',\n" .
       "  max_doc_id int(11) NOT NULL,\n" .
       "  PRIMARY KEY  (counter_id,`type`)\n" .
       ")";

print <<< CONFIGDONE
----------------------------------------------------------------------
An extra table "$incr_table" is required
----------------------------------------------------------------------

For supporting incremental updates for searching, an extra database
table is required. The definition of this table is:

$sql

Do you want to automatically create this table now?

CONFIGDONE;

while (TRUE)
{
    print "\nPlease enter \"yes\" or \"no\" > ";
    $yesno = trim(fgets(STDIN));
    if ($yesno == 'yes' || $yesno == 'no') { break; }
}

if ($yesno == 'yes') {
    phorum_db_interact(DB_RETURN_RES, $sql);
    print "\nTable \"$incr_table\" created.\n\n";
}


exit(0);
?>
